syntax = "proto3";

import "proto/transaction.proto";

package slog.internal;

message Batch {
    uint64 id = 1;
    repeated Transaction transactions = 2;
    // All txns in this batch has this same type.
    // Can either be SINGLE_HOME or MULTI_HOME. 
    // LOCK_ONLY txns are also considered SINGLE_HOME.
    TransactionType transaction_type = 3;
    // For recording the event time for all
    // transactions in this batch simultaneously
    repeated TransactionEventInfo events = 4;
}

message LocalBatchOrder {
    // generator is the machine that the 
    // batch is generated from
    uint32 generator = 1;
    uint32 slot = 2;
    uint32 leader = 3;
}

message RemoteBatchOrder {
    uint64 batch_id = 1;
    uint32 slot = 2;
    uint32 home = 3;
    bool need_ack = 4;
}

message Envelope {
    oneof type {
        Request request = 1;
        Response response = 2;
        bytes raw = 3;
    }
    uint32 from = 4;
}

/***********************************************
                    REQUESTS
***********************************************/
/**
 * The first message of a communication between two entities is
 * always a Request. Some types of Request needs a Response
 * (e.g. LookUpMasterRequest); names of these requests end with "Request".
 * Some others are one-way requests (e.g. PaxosPropose) and do not 
 * end with "Request".
 */
message Request {
    oneof type {
        Ping ping = 1;
        Signal signal = 2;
        BrokerRedirect broker_redirect = 3;
        ForwardTransaction forward_txn = 4;
        LookupMasterRequest lookup_master = 5;
        ForwardBatchData forward_batch_data = 6;
        ForwardBatchOrder forward_batch_order = 7;
        BatchReplicationAck batch_replication_ack = 8;
        PaxosPropose paxos_propose = 9;
        PaxosAcceptRequest paxos_accept = 10;
        PaxosCommitRequest paxos_commit = 11;
        RemoteReadResult remote_read_result = 12;
        FinishedSubtransaction finished_subtxn = 13;
        StatsRequest stats = 14;
        GraphLog graph_log = 15;
        /* Janus */
        JanusPreAcceptRequest janus_pre_accept = 16;
        JanusAcceptRequest janus_accept = 17;
        JanusCommit janus_commit = 18;
        JanusInquireRequest janus_inquire = 19;
    }
}

message Ping {
    int64 src_time = 1;
    uint32 dst = 2;
}

/**
 * Generic signal message
 */
message Signal {
}

message BrokerRedirect {
    uint64 tag = 1;
    uint32 channel = 2;
    bool stop = 3;
}

message ForwardTransaction {
    Transaction txn = 1;
}

message LookupMasterRequest {
    repeated uint64 txn_ids = 1;
    repeated bytes keys = 2;
}

message ForwardBatchData {
    repeated Batch batch_data = 1;
    // Machine that generated the batch
    uint32 generator = 2;
    // Batches generated by the same machine need to follow the
    // order of creation. This field is used to number the batches
    // following that order. It always start from 0 and increment by 1
    uint32 generator_position = 3;
}

message ForwardBatchOrder {
    oneof locality {
        LocalBatchOrder local_batch_order = 1;
        RemoteBatchOrder remote_batch_order = 2;
    }
}

message BatchReplicationAck {
    uint64 batch_id = 1;
}

message PaxosPropose {
    uint64 value = 1;
}

message PaxosAcceptRequest {
    uint32 ballot = 1;
    uint32 slot = 2;
    uint64 value = 3;
}

message PaxosCommitRequest {
    uint32 slot = 1;
    uint64 value = 2;
    uint32 leader = 3;
}

message RemoteReadResult {
    uint64 txn_id = 1;
    bool deadlocked = 2;
    uint32 partition = 3;
    repeated KeyValueEntry reads = 4;
    bool will_abort = 5;
    string abort_reason = 6;
    AbortCode abort_code = 7;
}

message FinishedSubtransaction {
    Transaction txn = 1;
    uint32 partition = 2;
}

message StatsRequest {
    uint64 id = 1;
    uint64 level = 2;
}

message GraphLogEntry {
    uint64 txn_id = 1;
    int32 num_partitions = 2;
    bool is_complete = 3;
    repeated uint64 incoming_edges = 4;
}

message GraphLog {
    repeated GraphLogEntry entries = 1;
}

message JanusDependency {
    uint64 txn_id = 1;
    uint32 target_partition = 2;
    uint64 participants_bitmap = 3;
}

message JanusPreAcceptRequest {
    Transaction txn = 1;
    int32 ballot = 2;
}

message JanusAcceptRequest {
    uint64 txn_id = 1;
    repeated JanusDependency deps = 2;
    int32 ballot = 3;
}

message JanusCommit {
    uint64 txn_id = 1;
    repeated JanusDependency deps = 2;
    Transaction txn = 3;
}

message JanusInquireRequest {
    repeated uint64 txn_id = 1;
}

/***********************************************
                    RESPONSES
***********************************************/
/**
 * A response is always preceeded by a Request
 */
message Response {
    oneof type {
        Pong pong = 1;
        LookupMasterResponse lookup_master = 2;
        PaxosAcceptResponse paxos_accept = 3;
        PaxosCommitResponse paxos_commit = 4;
        StatsResponse stats = 6;
        /* Janus */
        JanusPreAcceptResponse janus_pre_accept = 7;
        JanusAcceptResponse janus_accept = 8;
        JanusInquireResponse janus_inquire = 9;
    }
}

/**
 * For debugging and testing purposes
 */
message Pong {
    int64 src_time = 1;
    int64 dst_time = 2;
    uint32 dst = 3;
}

message KeyMasterMetadata {
    bytes key = 1;
    MasterMetadata metadata = 2;
}

message LookupMasterResponse {
    repeated uint64 txn_ids = 1;
    repeated KeyMasterMetadata lookup_results = 2;
}

message PaxosAcceptResponse {
    uint32 ballot = 1;
    uint32 slot = 2;
}

message PaxosCommitResponse {
    uint32 slot = 1;
}

message StatsResponse {
    uint64 id = 1;
    bytes stats_json = 2;
}

message JanusPreAcceptResponse {
    uint64 txn_id = 1;
    bool ok = 2;
    repeated JanusDependency deps = 3;
}

message JanusAcceptResponse {
    uint64 txn_id = 1;
    bool ok = 2;
}

message JanusInquiryResult {
    uint64 txn_id = 1;
    bool executed = 2;
    repeated JanusDependency deps = 3;
}

message JanusInquireResponse {
    repeated JanusInquiryResult results = 1;
}
